kvdb/postgres/readwrite_tx.go

```diff
@@ -0,0 +1,295 @@
+func (tx *readWriteTx) ReadWriteBucket(key []byte) walletdb.ReadWriteBucket {
+	table := tx.toTableName(key)
+
+	// If the key can't be mapped to a table name, open the bucket from the
+	// hex table.
+	if table == "" {
+		bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+		return bucket.NestedReadWriteBucket(key)
+	}
+
+	var value int
+	err := tx.tx.QueryRowContext(
+		context.TODO(),
+		"SELECT 1 FROM pg_catalog.pg_tables WHERE schemaname='public' and tablename=$1",
+		table,
+	).Scan(&value)
+
+	if err == sql.ErrNoRows {
+		return nil
+	}
+
+	return newReadWriteBucket(tx, table, nil)
+}
+
+func getCreateTableSql(table string) string {
+	return `
+CREATE TABLE IF NOT EXISTS public.` + table + `
+(
+    key bytea NOT NULL,
+    value bytea,
+    parent_id bigint,
+    id bigserial PRIMARY KEY,
+    sequence bigint,
+    CONSTRAINT ` + table + `_parent FOREIGN KEY (parent_id)
+        REFERENCES public.` + table + ` (id) MATCH SIMPLE
+        ON UPDATE NO ACTION
+        ON DELETE CASCADE
+);
+
+CREATE INDEX IF NOT EXISTS ` + table + `_p
+    ON public.` + table + ` USING btree
+    (parent_id ASC NULLS LAST);
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_up
+    ON public.` + table + ` USING btree
+    (parent_id ASC NULLS LAST, key ASC NULLS LAST) WHERE parent_id IS NOT NULL;
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_unp ON public.` + table + ` (key) WHERE parent_id IS NULL;
+`
+}
+
+// CreateTopLevelBucket creates the top level bucket for a key if it
+// does not exist.  The newly-created bucket it returned.
+func (tx *readWriteTx) CreateTopLevelBucket(key []byte) (walletdb.ReadWriteBucket, error) {
+	table := tx.toTableName(key)
+
+	if table == "" {
+		bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+		return bucket.CreateBucketIfNotExists(key)
+	}
+
+	createTableSql := getCreateTableSql(table)
+	_, err := tx.tx.ExecContext(context.TODO(), createTableSql)
+	if err != nil {
+		return nil, err
+	}
+
+	return newReadWriteBucket(tx, table, nil), nil
+}
+
+// DeleteTopLevelBucket deletes the top level bucket for a key.  This
+// errors if the bucket can not be found or the key keys a single value
+// instead of a bucket.
+func (tx *readWriteTx) DeleteTopLevelBucket(key []byte) error {
+	bucket := tx.ReadWriteBucket(key)
+	if bucket == nil {
+		return walletdb.ErrBucketNotFound
+	}
+
+	table := bucket.(*readWriteBucket).table
+
+	_, err := tx.tx.ExecContext(context.TODO(), "DROP TABLE public."+table+";")
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.tx.ExecContext(context.TODO(), "DELETE FROM "+tx.db.sequenceTableName+" WHERE table_name=$1", table)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Commit commits the transaction if not already committed.
+func (tx *readWriteTx) Commit() error {
+	// Commit will fail if the transaction is already committed.
+	if !tx.active {
+		return walletdb.ErrTxClosed
+	}
+
+	// Try committing the transaction.
+	if err := tx.tx.Commit(); err != nil {
+		return err
```

Comment:
Similarly to previous comment, I believe this can deadlock if the commit fails.

Reply:
Fixed too